WIP: Output Tracking API

vanzin opened the pull request:
This is a very hacky P.O.C. of an output tracking API so that shuffle I/O plugins can better hook up to the Spark scheduler.

The idea here is that Spark's existing `MapOutputTracker` is not a good way of tracking map outputs when the underlying shuffler storage is not the default one. In spite of the Spark-isms of the current write API, plugins have the ability to do other things like coalesce outputs to make reading more efficient, or move blocks around to different locations, and the existing output tracker has no notion of that.

The basic change is to have `MapOutputTracker` basically be a proxy between the scheduler and the plugin; the main thing it handles internally is the distribution of shuffle metadata so that executors can get that metadata when reading shuffle data. Everything else (like the equivalent of `MapStatus`) is completely opaque to Spark, and is handled by the plugins.

As I mentioned, it's super hacky currently. In particular, there is still a lot of baggage from the existing shuffle code on the read path. And I didn't implement error handling at all, although I started with some basic API. To try to itemize the problems and missing things:

- the internal shuffle read code in Spark depends on a lot of internal state that doesn't make much sense to expose to plugins. For example, things like serializer configuration, which is encoded in the shuffle dependency internally, and block IDs, which are also used when making decisions about how to read data (although this one should be simpler to fix).

- error handling is not implemented; the use and handling of fetch failure issues needs to be properly abstracted so that the scheduler can notify plugins when reading fails, and provide the right metadata to plugins so they can invalidate or fix their internal state.

- to support async updates of plugin state, we need that to either go through the `MapOutputTracker`, or have some callback that plugins can call to invalidate the serialization cache; another option is to leave the distribution of that data to plugins themselves (maybe with some helper code provided by Spark).

- LocalShuffleRowRDD uses some very specialized APIs to get shuffle data which I haven't really bothered to investigate at this point. (I've commented most of that code out in this branch.)

I took some inspiration from Ilan's PR (https://github.com/apache/spark/pull/26772) in some places; that can be seen mostly in `LocalDiskShuffleExecutorComponents` and things mentioned in comments there. That's the main part of the pain of abstracting the read side.

(Ilan's PR tries to follow what the write side does, and basically encode the way the existing shuffle manager reads shuffle blocks one at a time. It may be a feasible approach, but I still think that we need to solve the metadata problem too, and the error handling, which I didn't really see if Ilan's PR solves.)

Commit:
WIP.

TODO:
- refactor map output tracker
  - move tracking code into plugin
  - delegate calls to the plugin
- fix DAGScheduler to use MapOutputMetadata

Commit:
WIP2: hacking map output tracker.

Commit:
Hack hack hack the read side.

TODO:
- ShuffleBlockFetcherIterator needs to be broken down.
  - First part does only block fetching.
  - Second part does stuff that needs internal Spark stuff,
    like checking for corruption (which needs the serializer
    stuff).
- The block fetching part should implement the (Java) plugin API.
- Locking should be in the MapOutputTracker to avoid racing reads
  during serialization.

Commit:
WIP: move locking to shared code.

Commit:
Remove unused code.

hiboyang commented:
+1 for the opaque MapOutputMetadata and make MapStatus extend MapOutputMetadata!

Commit:
Hacky single-mapper read API.

Commit:
Get rid of getStatistics().

Doesn't seem to be used for anything.

Commit:
First stab at preferred locs API.

Commit:
Revert "Get rid of getStatistics()."

This reverts commit fbe7fe3c80a1e0e4bf49220c8424d6d3de1aec70.

Commit:
Small fix.

vanzin commented:
Current state: I've abstracted a few more things into the API, although it may not look great. Only `MapOutputTracker.getStatistics` remains, and while I could hack an API for that, it raises some questions about the relationship between disaggregated shuffle storage and adaptive query execution (which AFAICT is the only code that uses that).

Commit:
Crude fetch failure handling API, no implementation yet.

Commit:
More work on error handling abstraction.

Not really hooked up to the scheduler yet, need to modify fetch failed exception
for this to work.

Commit:
Need map index in block metadata.

Commit:
Trying to hook up fetch failures to the shuffle plugin.

Commit:
Add bm address to failure metadata and use it in internal error handling paths.

This is very hacky since it hardcodes error handling for the internal
plugin in a few cases (e.g. removing executors on fetch failures). But
BlockManagerId is a (primarily) internal concept and I'm trying to avoid
exposing in this API.

gcz2022 commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -536,6 +536,9 @@ class SparkContext(config: SparkConf) extends Logging {
     }
 
+    env.mapOutputTracker.asInstanceOf[MapOutputTrackerMaster].outputTracker =
+      Option(_shuffleDriverComponents.createOutputTracker().orElse(null))
```
What if a ShuffleManager plugin wants to register its customized `MapOutputTracker` to driver? This seems only enable shuffle I/O plugin to hook in driver.

gcz2022 replied:
It can be not complex to support a ShuffleManager plugin here : P : when `spark.shuffle.manager` is not default value,  
`xxx.outputTracker = _env.shuffleManager.createOutputTracker`?

vanzin replied:
ShuffleManager is not a public API, which is one of the reasons why the new API is being created.

If you'd like to discuss making ShuffleManager a proper public API, then it's a separate discussion, and we'd kinda have two different approaches to shuffle pluggability. I'd rather we as a community focus on just one, and make that one cover as much ground as possible.

gcz2022 replied:
Thanks for your explanation!

I see ShuffleManager is a private[spark] API, but still we dynamically instantiate a ShuffleManager implementation specified by "spark.shuffle.manager" while creating SparkEnv. This means Spark has been using it like a public API for long time(so many people have been using it).

I understand your view on focusing just one pluggable API at a time. But my concerns are that, if we only support Shuffle I/O API in this new map output tracking design, developers may not benefit under these conditions/reasons: 1) The already-implemented ShuffleManagers cannot take use of this, unless they port the implementation to work under Shuffle I/O API 2) In some cases, the shuffle algorithms may need slight changes from default sort-based shuffle, so they cannot implement a Shuffle I/O to fullfill their needs 3) Develop a Shuffle I/O plugin is harder than ShuffleManager and needs more learning, some developers may still stick to the original ShuffleManager API

vanzin replied:
> but still we dynamically instantiate a ShuffleManager implementation specified by "spark.shuffle.manager" while creating SparkEnv

That's historical from when Spark had a different manager (the hash shuffle manager). It was never meant for others to extend, because this API exposes way too many internal Spark things.

I understand your desire to try to make that use case official. But if we go that route, we should revisit the whole idea of abstracting the shuffle I/O instead of the shuffle manager. Having the two is way too much work.

gcz2022 replied:
Thank you : ) I guess if so we need to make the new API really general, applying to as more cases as possible.

mccheah replied:
Though I think it's correct to treat `ShuffleManager` as not being meant to be extended, unfortunately that's already being done in a bunch of places - the Splash shuffle manager comes to mind. We can hack around this here by checking `instanceOf` or something similar.

gcz2022 replied:
@mccheah ^_^ +1

itskals commented on core/src/main/java/org/apache/spark/shuffle/api/ShuffleMapOutputWriter.java:
```diff
@@ -63,7 +63,7 @@
    * The returned array should contain, for each partition from (0) to (numPartitions - 1), the
    * number of bytes written by the partition writer for that partition id.
```
May be this comment needs some update for the changed interface.

vanzin replied:
This is a very hacky P.O.C. Lots of things need update. Tests don't even compile.

mccheah commented on core/src/main/java/org/apache/spark/shuffle/api/ShuffleExecutorComponents.java:
```diff
@@ -73,4 +74,13 @@ ShuffleMapOutputWriter createMapOutputWriter(
+
+  // Return an iterator for the shuffle data defined by the args.
+  Iterator<ShuffleInputStream> readShuffle(
+      int shuffleId,
```
In our proposed version of the API we provide the specific list of blocks to fetch, which I think is easier to consume rather than a range: https://github.com/palantir/spark/blob/master/core/src/main/java/org/apache/spark/shuffle/api/ShuffleExecutorComponents.java#L63

mccheah replied:
Although that would require passing in the metadata per block instead of sending a single `ShuffleMetadata` object. I presume the `ShuffleMetadata` object would need to have functionality to retrieve per-block metadata, although it's up to the implementation to provide that.

mccheah commented on core/src/main/java/org/apache/spark/shuffle/api/ShuffleOutputTracker.java:
```diff
@@ -0,0 +1,69 @@
+  void registerShuffle(int shuffleId, int numMaps);
+
+  // Called for each task (from DAGScheduler)
+  boolean registerOutput(int shuffleId, int mapIndex, MapOutputMetadata metadata);
```
Should we use `mapId` everywhere instead of `mapIndex`?

mccheah reviewed (commented):
This looks very solid. I took a closer pass through the code, and so far I don't really have major concerns. The only thing is that the reader API might be better served by giving it the list of blocks to fetch instead of a range of start-to-end partitions.

vanzin commented:
HI guys, thanks for the comments here, but let me add some context.

The main thing is that I won't really have bandwidth to move this forward. Sorry I haven't been able to make it to the calls, but that's also a part of the "don't have bandwidth right now" issue.

So if you guys think this is a good approach, feel free to use it as inspiration or even borrow the code (although let me say again how hacky it is).

If you want to follow this approach these are the things I think you should keep in mind:

- The read API I created is asymmetrical to the write API. This will at least makes things a little odd when you start thinking about compression and encryption, which currently are handled by Spark itself instead of being handled by the plugins. That needs to be properly thought out to make sure it works for plugins.

- I did not add any API here to allow plugins to update map locations. Maybe it's not needed and plugins can do it internally by using their own RPC? (https://github.com/apache/spark/pull/26670 should make that a little easier, but nobody has really reviewed it.) The output tracker needs at least to know that the metadata has changed, somehow.

The latter would probably require some API that exposes Spark functionality to plugins (instead of the other way around). Similar to `PluginContext` in the plugin API.
